alf/config_util.py

```diff
@@ -112,19 +112,18 @@ def func(self, a, b):
             immutable value to an existing immutable value.
         raise_if_used (bool): If True, ValueError will be raised if trying to
             config a value which has already been used.
-        sole_init (bool): If True, the config value can only be set once. Any
-            previous or future calls will raise a ValueError. This is helpful
-            in enforcing a singular point of initialization, thus eliminating
-            any potential side effects from possible prior or future overrides.
-            This flag overrides the mutable flag if True. For users wanting this
-            to be the default behavior, the ALF_SOLE_CONFIG env variable can be
-            set to 1.
-        override (bool): If True, the value of the config will be set regardless
-            of any pre-existing ``mutable`` or ``sole_init`` settings. This should
-            be used only when absolutely necessary (e.g., a teacher-student training
-            loop, where the student must override certain configs inherited from the
-            teacher). Otherwise, use ``mutable`` or ``sole_init`` instead. If override
-            is True, the config's ``mutable`` or ``sole_init`` values are not changed.
+        sole_init (bool): If True, the config value can no longer be set again
+            after this config call. Any future calls will raise a RuntimeError.
+            This is helpful in enforcing a singular point of initialization,
+            thus eliminating any potential side effects from possible future
+            overrides. For users wanting this to be the default behavior, the
+            ALF_SOLE_CONFIG env variable can be set to 1.
+        override_sole_init (bool): If True, the value of the config will be set
+            regardless of any previous ``sole_init`` setting. This should be used
+            only when absolutely necessary (e.g., a teacher-student training loop,
+            where the student must override certain configs inherited from the
+            teacher). If the config is immutable, a warning will be declared with
+            no changes made.
```

Comment:
instead of quietly ignoring the config, should we just throw error for this case?

Reply:
So the reason we cannot do this is because `alf.pre_config` and `--conf_param` set values with `mutable=False` (that is how those values avoid getting overwritten). Therefore, there may be times we want to override our conf files using one of the two options and raising an error in this situation is problematic. A workaround for this was to possibly introduce a `pre_configged` state variable for configs, but I thought that may be unnecessary. Let me know what you think.

Reply:
I see.  You are right.  Let's go with current way for the PR.

Separately, are there some examples of these cases?  It would be good to understand a bit more, and see if we can design better solutions.